packages/r2/src/r2Object.ts

```diff
@@ -0,0 +1,323 @@
+  httpMetadata?: R2HTTPMetadata | Headers
+): R2HTTPMetadata {
+  if (httpMetadata === undefined) return {};
+  if (httpMetadata instanceof Headers) {
+    const cExpiry = httpMetadata.get("cache-expiry");
+    return {
+      contentType: httpMetadata.get("content-type") ?? undefined,
+      contentLanguage: httpMetadata.get("content-language") ?? undefined,
+      contentDisposition: httpMetadata.get("content-disposition") ?? undefined,
+      contentEncoding: httpMetadata.get("content-encoding") ?? undefined,
+      cacheControl: httpMetadata.get("cache-control") ?? undefined,
+      cacheExpiry: cExpiry ? new Date(cExpiry) : undefined,
+    };
+  } else {
+    // remove variables that are not part of the HTTP metadata
+    const httpMetadataList = [
+      "contentType",
+      "contentLanguage",
+      "contentDisposition",
+      "contentEncoding",
+      "cacheControl",
+      "cacheExpiry",
+    ];
+    for (const key of Object.keys(httpMetadata)) {
+      if (!httpMetadataList.includes(key)) {
+        delete httpMetadata[key as keyof R2HTTPMetadata];
+      }
+    }
+
+    return httpMetadata;
+  }
+}
+
+// false -> the condition testing "failed"
+export function testR2Conditional(
+  conditional: R2Conditional,
+  metadata?: R2ObjectMetadata
+): boolean {
+  const { etagMatches, etagDoesNotMatch, uploadedBefore, uploadedAfter } =
+    conditional;
+  // If the object doesn't exist
+  if (metadata === undefined) {
+    // the etagDoesNotMatch and uploadedBefore automatically pass
+    // etagMatches and uploadedAfter automatically fail if they exist
+    if (etagMatches !== undefined || uploadedAfter !== undefined) {
+      return false;
+    } else {
+      return true;
+    }
+  }
+
+  const { etag, uploaded } = metadata;
+
+  // ifMatch check
+  const ifMatch = etagMatches ? matchStrings(etagMatches, etag) : null;
+  if (ifMatch === false) return false;
+
+  // ifNoMatch check
+  const ifNoneMatch = etagDoesNotMatch
+    ? !matchStrings(etagDoesNotMatch, etag)
+    : null;
+  if (ifNoneMatch === false) return false;
+
+  // ifUnmodifiedSince check
+  if (
+    ifMatch !== true && // if "ifMatch" is true, we ignore date checking
+    uploadedBefore !== undefined &&
+    uploaded > uploadedBefore
+  ) {
+    return false;
+  }
+
+  // ifModifiedSince check
+  if (
+    ifNoneMatch !== true && // if "ifNoneMatch" is true, we ignore date checking
+    uploadedAfter !== undefined &&
+    uploaded < uploadedAfter
+  ) {
+    return false;
+  }
+
+  return true;
+}
+
+function matchStrings(a: string | string[], b: string): boolean {
+  if (typeof a === "string") return a === b;
+  else return a.includes(b);
+}
+
+// headers can be a list: e.g. ["if-match", "a, b, c"] -> "if-match: [a, b, c]"
+function parseHeaderArray(input?: string): undefined | string | string[] {
+  if (input === undefined) return;
+  if (typeof input !== "string") return;
+  const list = input.split(",");
+  if (list.length === 1) return list[0];
```

Comment:
I couldn't create a test case where this would be true.

The first case is verifying that the `,` didn't exist, so the string wouldn't be modified, in other words, if the user added a space at the end of their singular test, I think it shouldn't be modified.

The second case is trimming the spacing between the commas, so 'a, b, c' should remove the spaces after the commas.

If you can think of an example that trim would be needed feel free to push back.

On second thought, I'm goign to change the function to be more clear:
```ts
// headers can be a list: e.g. ["if-match", "a, b, c"] -> "if-match: [a, b, c]"
function parseHeaderArray(input?: string): undefined | string | string[] {
  if (typeof input !== "string") return;
  if (!input.includes(",")) return input;
  const list = input.split(",");
  return list.map((x) => x.trim());
}
```